README.rst
===================
kotti_content_proxy
===================

This is an extension to the Kotti CMS that allows you to add *content proxies* to your Kotti site.

A *content proxy* is very similar to a *symbolic link* on UNIX systems.
It allows you to refer to a single content node from multiple places in your node tree without having to duplicate the content itself.
In other words: it will show up in the navigation in multiple places, but still only exist once.

For the moment only the ``view`` view and the ``__acl__`` attribute are proxied.
This means that you can only see proxies of proxied objects that you are allowed to see; when you request the "view" of the ContentProxy object it will return the rendered view of the proxied object instead (of course also respecting its permissions).

For future versions it is planned to also proxy as many object attributes as feasible / sensible.

`Find out more about Kotti`_

Setup
=====

To activate the kotti_content_proxy add-on in your Kotti site, you need to add an entry to the ``kotti.configurators`` setting in your Paste Deploy config.
If you don't have a ``kotti.configurators`` option, add one.
The line in your ``[app:main]`` section could then look like this::

  kotti.configurators = kotti_content_proxy.kotti_configure

With this, you'll be able to add content proxy items in your site.

Todo
====

- Make selection of the proxied object via the UI more user friendly.
  Maybe use ``kotti_tinymce``'s ``kottibrowser``.

- Proxy everything that makes sense.

.. _Find out more about Kotti: http://pypi.python.org/pypi/Kotti

Authors
=======

- Andreas Kaiser (disko).
- Piotr Dobrogost (pdobrogost)


CHANGES.rst
Changelog
=========

0.5.0 - unreleased
------------------

- No changes yet.

0.4.1 - 2016-01-11
------------------

- Fix broken packaging.  Sorry for the inconvenience!

0.4.0 - 2015-10-07
------------------

- Now we take object to be proxied instead of its id.
  **This is a backward incompatible change.**

0.3.0 - 2015-10-05
------------------

- Names of proxied attributes are now kept in class attribute and read from
  ``kotti_content_proxy.proxied_attributes`` setting.

0.2.0 - 2015-09-30
------------------

- Proxy now declares unique ``polymorphic_identity`` for use in SQLAlchemy.
- Added proxying of custom attributes.

0.1.0
-----

- Initial release.


kotti_content_proxy/tests/conftest.py
# -*- coding: utf-8 -*-

from pytest import fixture

pytest_plugins = "kotti"


@fixture
def foo_proxy(root, db_session):

    from kotti.resources import Document
    from kotti_content_proxy.resources import ContentProxy

    foo = root[u'foo'] = Document(title=u'Foo', description=u'Bar',
                                  body=u'This is the Foo document')
    db_session.flush()

    # create a proxy for the 'foo'
    proxy = root[u'proxy'] = ContentProxy(proxied_object=foo, title=u'Proxy',
                                          description=u'This is a proxy')
    db_session.flush()

    return dict(
        foo=foo,
        proxy=proxy,
    )


kotti_content_proxy/tests/test_views.py
# -*- coding: utf-8 -*-


def test_view_view(foo_proxy, webtest):

    foo = webtest.get('/foo')
    proxy = webtest.get('/proxy')


kotti_content_proxy/tests/test_models.py
# -*- coding: utf-8 -*-

import pytest


def test_model(foo_proxy):
    assert foo_proxy[u'proxy'].proxied_object == foo_proxy[u'foo']


def test_proxying_custom_attributes(foo_proxy):
    import kotti_content_proxy

    mock_settings = {
        'kotti.available_types': '',
        'pyramid.includes': '',
    }

    settings = {
        'kotti_content_proxy.proxied_attributes':
            'some_attribute another_attribute'
    }

    settings.update(mock_settings)

    kotti_content_proxy.kotti_configure(settings)

    foo = foo_proxy['foo']
    foo.some_attribute = u'some value'
    proxy = foo_proxy['proxy']

    assert proxy.some_attribute == foo.some_attribute
    with pytest.raises(AttributeError) as excinfo:
        proxy.another_attribute
    assert excinfo.value.args[0] == "'Document' object has no attribute 'another_attribute'"  # noqa


kotti_content_proxy/views.py
# -*- coding: utf-8 -*-

"""
Created on 2014-09-23
:author: Andreas Kaiser (disko)
"""

import colander

from kotti.resources import Node
from kotti.util import render_view
from kotti.views.form import AddFormView
from kotti.views.form import EditFormView

from pyramid.response import Response
from pyramid.view import view_config
from pyramid.view import view_defaults

from kotti_content_proxy import _
from kotti_content_proxy.resources import ContentProxy


class ContentProxySchema(colander.MappingSchema):
    """ ContentProxy add / edit schema """

    title = colander.SchemaNode(
        colander.String(),
        title=_(u'Title'))
    proxied_id = colander.SchemaNode(
        colander.Integer(),
        title=_(u'Proxied ID'))


@view_config(name=ContentProxy.type_info.add_view, permission='add',
             renderer='kotti:templates/edit/node.pt')
class ContentProxyAddForm(AddFormView):
    """ ContentProxy add view """

    schema_factory = ContentProxySchema
    add = ContentProxy
    item_type = _(u"ContentProxy")

    def save_success(self, appstruct):
        proxied_id = appstruct.pop('proxied_id')
        proxied_object = Node.query.filter(Node.id == proxied_id).one()
        appstruct['proxied_object'] = proxied_object
        return super(ContentProxyAddForm, self).save_success(appstruct)


@view_config(context=ContentProxy, name='edit', permission='edit',
             renderer='kotti:templates/edit/node.pt')
class ContentProxyEditForm(EditFormView):
    """ ContentProxy edit view """

    schema_factory = ContentProxySchema


@view_defaults(context=ContentProxy, permission='view')
class ContentProxyView(object):
    """ ContentProxy view(s) """

    def __init__(self, context, request):
        """ Constructor

        :param context: The proxy object
        :type context: :class:`kotti_content_proxy.resources.ContentProxy`

        :param request: The current request
        :type request: :class:`pyramid.request.Request`
        """

        super(ContentProxyView, self).__init__()
        self.context = context
        self.request = request

    @view_config(name='view')
    def view(self):
        """ Rendered ``view`` view of the proxied object.

        :result: Rendered response
        :rtype: :class:`pyramid.response.Response`
        """

        return Response(
            render_view(
                self.context.proxied_object,
                self.request,
                name='view',
                secure=True))


kotti_content_proxy/resources.py
# -*- coding: utf-8 -*-

"""
Created on 2014-09-23
:author: Andreas Kaiser (disko)
"""

from kotti.resources import Content
from sqlalchemy import Column
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy.orm import relationship

from kotti_content_proxy import _


class ContentProxy(Content):
    """ A ContentProxy refers to any other content node in the site. """

    id = Column(
        Integer,
        ForeignKey('contents.id'),
        primary_key=True)
    proxied_id = Column(
        Integer,
        ForeignKey('contents.id'),
        nullable=False)

    proxied_object = relationship(Content, foreign_keys=[proxied_id])

    # This can't be None as __getattribute__ is being called by SQLAlchemy's
    # instrumentation before __init__ is called. See
    # sqlalchemy.orm.instrumentation.ClassManager._new_state_if_none()
    proxied_attrs = set()
    _default_proxied_attrs = set(('__acl__',))

    __mapper_args__ = Content.__mapper_args__.copy()
    __mapper_args__.update({
        'inherit_condition': (id == Content.id),
        'polymorphic_identity': 'content_proxy',
    })

    type_info = Content.type_info.copy(
        name=u'ContentProxy',
        title=_(u'Content Proxy'),
        add_view=u'add_content_proxy',
        addable_to=[u'Document'],
        )

    def __init__(self, proxied_object=None, **kwargs):
        """ Constructor

        :param proxied_object: proxied object
        :type proxied_object: :class:`kotti.resources.Content`

        :param **kwargs: see :class:`kotti.resources.Content`
        :type **kwargs: variant
        """

        super(ContentProxy, self).__init__(**kwargs)
        self.proxied_object = proxied_object

    def __getattribute__(self, key):
        """ Proxy some attributes.
            See https://www.inkling.com/read/learning-python-mark-lutz-4th/chapter-37/--getattr---and---getattribute--  # noqa
        """
        super_getattribute = super(ContentProxy, self).__getattribute__
        proxied_attrs = super_getattribute('proxied_attrs')

        if key in proxied_attrs:
            return self.proxied_object.__getattribute__(key)

        return super(ContentProxy, self).__getattribute__(key)
